File: apps/server/dekobloko_server/accounts.py
from __future__ import annotations

from dataclasses import dataclass
import hashlib
import json
from pathlib import Path


@dataclass(frozen=True)
class AuthResult:
    ok: bool
    display_name: str
    reason: str


class AccountStore:
    def __init__(self, path: Path, auto_register: bool) -> None:
        self.path = path
        self.auto_register = auto_register
        self.accounts: dict[str, dict[str, str]] = {}
        if path.exists():
            with path.open("r", encoding="utf-8") as handle:
                loaded = json.load(handle)
            self.accounts = dict(loaded.get("accounts", {}))

    def authenticate(self, username: str, password: str) -> AuthResult:
        normalized = self._normalize(username)
        existing = self.accounts.get(normalized)
        if existing is None:
            if not self.auto_register:
                return AuthResult(False, username, "unknown account")
            self.accounts[normalized] = {
                "display_name": username or "Player",
                "password_sha256": self._password_hash(password),
            }
            self._save()
            return AuthResult(True, username or "Player", "auto-registered")

        if existing.get("password_sha256") != self._password_hash(password):
            return AuthResult(False, existing.get("display_name", username), "bad password")
        return AuthResult(True, existing.get("display_name", username), "ok")

    def _save(self) -> None:
        self.path.parent.mkdir(parents=True, exist_ok=True)
        with self.path.open("w", encoding="utf-8") as handle:
            json.dump({"accounts": self.accounts}, handle, indent=2, sort_keys=True)
            handle.write("\n")

    def player_id(self, username: str) -> int:
        """Stable per-account id sent to the client at login.

        This used to be a hardcoded 1 for every account, which caused real
        identity corruption: the client stores the id and sends it back in the
        USERNAME SLOT on every reconnect. Base37-decoding 1 yields "A", so any
        player who reconnected -- which happens on every return-to-main-menu --
        silently became account "a" and had their scores filed there.

        Deriving it from the account name makes the value round-trip to the
        right player, and is deterministic so it survives a restart without
        needing to be persisted.

        The offset keeps ids clear of the small integers a stray field might
        hold. It is NOT a guarantee of non-collision with a real packed name --
        username_for_player_id() only matches ids of accounts that exist, which
        is what actually makes the lookup safe.
        """
        normalized = self._normalize(username)
        digest = hashlib.sha256(normalized.encode("utf-8")).digest()
        return 0x1000_0000 | (int.from_bytes(digest[:4], "big") & 0x0FFF_FFFF)

    def username_for_player_id(self, player_id: int) -> str | None:
        """Reverse player_id() -- which account was this id issued to?

        Returns None when the value is not a known account's id, in which case
        the caller should treat the slot as a packed username as before.
        """
        if player_id <= 0:
            return None
        for name in self.accounts:
            if self.player_id(name) == player_id:
                return name
        return None

    @staticmethod
    def _normalize(username: str) -> str:
        return username.strip().lower()

    @staticmethod
    def _password_hash(password: str) -> str:
        return hashlib.sha256(password.encode("utf-8")).hexdigest()

File: apps/server-js/src/accounts.js
'use strict';

// Port of dekobloko_server/accounts.py.

const fs = require('fs');
const path = require('path');
const { createHash } = require('crypto');

class AuthResult {
  // Mirrors the frozen dataclass AuthResult(ok, display_name, reason).
  constructor(ok, display_name, reason) {
    this.ok = ok;
    this.display_name = display_name;
    this.reason = reason;
    Object.freeze(this);
  }
}

function normalize(username) {
  return username.trim().toLowerCase();
}

function passwordHash(password) {
  return createHash('sha256').update(password, 'utf8').digest('hex');
}

class AccountStore {
  constructor(path_, auto_register) {
    this.path = path_;
    this.auto_register = auto_register;
    this.accounts = {};
    if (fs.existsSync(path_)) {
      const loaded = JSON.parse(fs.readFileSync(path_, 'utf8'));
      this.accounts = Object.assign({}, loaded.accounts || {});
    }
  }

  authenticate(username, password) {
    const normalized = normalize(username);
    const existing = this.accounts[normalized];
    if (existing === undefined) {
      if (!this.auto_register) {
        return new AuthResult(false, username, "unknown account");
      }
      this.accounts[normalized] = {
        display_name: username || "Player",
        password_sha256: passwordHash(password),
      };
      this._save();
      return new AuthResult(true, username || "Player", "auto-registered");
    }
    if (existing.password_sha256 !== passwordHash(password)) {
      return new AuthResult(
        false, existing.display_name !== undefined ? existing.display_name : username, "bad password");
    }
    return new AuthResult(
      true, existing.display_name !== undefined ? existing.display_name : username, "ok");
  }

  _save() {
    fs.mkdirSync(path.dirname(this.path), { recursive: true });
    // json.dump(..., sort_keys=True) parity: sorted account keys.
    const sortedAccounts = {};
    for (const key of Object.keys(this.accounts).sort()) {
      sortedAccounts[key] = this.accounts[key];
    }
    fs.writeFileSync(this.path, JSON.stringify({ accounts: sortedAccounts }, null, 2) + '\n');
  }

  // Stable per-account id sent to the client at login. Derived from the name
  // so it survives restarts; see accounts.py for the reconnect identity story.
  player_id(username) {
    const normalized = normalize(username);
    const digest = createHash('sha256').update(normalized, 'utf8').digest();
    return (0x10000000 + (digest.readUInt32BE(0) & 0x0fffffff)) >>> 0;
  }

  username_for_player_id(player_id) {
    if (!(player_id > 0)) return null;
    for (const name of Object.keys(this.accounts)) {
      if (this.player_id(name) === player_id) return name;
    }
    return null;
  }
}

module.exports = { AuthResult, AccountStore, normalize, passwordHash };
